File: collectors/reddit.py
"""Reddit collector using PRAW."""
from typing import List
import praw
from .base import BaseCollector, NewsItem


class RedditCollector(BaseCollector):
    """Fetch hot posts from tech subreddits."""
    
    def __init__(self, config: dict):
        super().__init__(config)
        self.reddit = None
        if self.enabled and config.get("client_id"):
            try:
                self.reddit = praw.Reddit(
                    client_id=config["client_id"],
                    client_secret=config["client_secret"],
                    user_agent=config.get("user_agent", "SocialHotspotBot/1.0"),
                )
            except Exception as e:
                print(f"[Reddit] Auth failed: {e}")
    
    def is_available(self) -> bool:
        return self.enabled and self.reddit is not None
    
    def collect(self) -> List[NewsItem]:
        if not self.is_available():
            return []
        
        items = []
        subreddits = self.config.get("subreddits", ["technology"])
        limit = self.config.get("limit", 15)
        
        for sub_name in subreddits:
            try:
                sub = self.reddit.subreddit(sub_name)
                for post in sub.hot(limit=limit):
                    if post.stickied:
                        continue
                    
                    hot_score = post.score * 0.6 + post.num_comments * 0.4
                    
                    items.append(NewsItem(
                        title=post.title,
                        url=post.url,
                        source="reddit",
                        source_name=f"r/{sub_name}",
                        hot_score=round(hot_score, 1),
                        category="technology",
                        language="en",
                        raw_data={
                            "score": post.score,
                            "comments": post.num_comments,
                            "upvote_ratio": post.upvote_ratio,
                        }
                    ))
            except Exception as e:
                print(f"[Reddit] Error in r/{sub_name}: {e}")
        
        return items

File: lib/collectors/reddit.ts
import type { Collector } from "./base";
import { createNewsItem } from "./base";

const DEFAULT_SUBREDDITS = [
  "technology",
  "programming",
  "MachineLearning",
  "science",
  "Futurology",
];

export class RedditCollector implements Collector {
  name = "reddit";
  private config: Record<string, unknown>;
  private snoowrap: any = null;

  constructor(config: Record<string, unknown>) {
    this.config = config;
    const clientId = config.client_id as string | undefined;
    const clientSecret = config.client_secret as string | undefined;

    if (clientId && clientSecret) {
      try {
        // Dynamic import to avoid crash if package not installed
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const snoowrap = require("snoowrap");
        this.snoowrap = new snoowrap({
          clientId,
          clientSecret,
          userAgent: (config.user_agent as string) ?? "SocialHotspotBot/1.0",
        });
      } catch (err) {
        console.error(`[Reddit] Auth failed: ${err}`);
      }
    }
  }

  async collect() {
    if (!this.snoowrap) return [];

    const items = [];
    const subreddits = (this.config.subreddits as string[]) ?? DEFAULT_SUBREDDITS;
    const limit = (this.config.limit as number) ?? 25;

    for (const subName of subreddits) {
      try {
        const sub = await this.snoowrap.getSubreddit(subName);
        const posts = await sub.getHot({ limit });

        for (const post of posts) {
          if (post.stickied) continue;

          const hotScore = post.score * 0.6 + post.num_comments * 0.4;

          items.push(
            createNewsItem(
              post.title,
              post.url,
              "reddit",
              `r/${subName}`,
              {
                hot_score: Math.round(hotScore * 10) / 10,
                category: "technology",
                language: "en",
                raw_data: {
                  score: post.score,
                  comments: post.num_comments,
                  upvote_ratio: post.upvote_ratio,
                },
              }
            )
          );
        }
      } catch (err) {
        console.error(`[Reddit] Error in r/${subName}: ${err}`);
      }
    }

    return items;
  }
}
